src/dev/run_prepush_hook.ts

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { existsSync } from 'node:fs';
+import { simpleGit } from 'simple-git';
+import cliProgress from 'cli-progress';
+import chalk from 'chalk';
+import Table from 'cli-table3';
+import { run } from '@kbn/dev-cli-runner';
+import { REPO_ROOT } from '@kbn/repo-info';
+import { ToolingLog } from '@kbn/tooling-log';
+import { getTimeReporter } from '@kbn/ci-stats-reporter';
+
+import { File } from './file';
+import { checkFileCasing } from './preflight_check/check_file_casing';
+import { checkTypescriptFiles } from './preflight_check/run_tsc';
+import { runUnitTests } from './preflight_check/run_unit_tests';
+import { esLintFiles } from './preflight_check/run_eslint';
+
+const HASH_FOR_NULL_OR_DELETED_FILE = '00000000000';
+
+function getDefaults(taskName: string) {
+  const TASK_COL_LENGTH = 10;
+  return {
+    task: chalk.blue(taskName.padEnd(TASK_COL_LENGTH)),
+    filename: '',
+  };
+}
+
+function nonNullable<T>(value: T): value is NonNullable<T> {
+  return value != null;
+}
+
+run(
+  async ({ log, flags }) => {
+    const toolingLog = new ToolingLog({
+      level: 'info',
+      writeTo: process.stdout,
+    });
+
+    const runStartTime = Date.now();
+    const reportTime = getTimeReporter(toolingLog, 'scripts/run_prepush_hook');
+
+    process.env.IS_KIBANA_PREPUSH_HOOK = 'true';
+
+    const git = simpleGit(REPO_ROOT);
+
+    if (!(await git.branchLocal().status()).isClean()) {
```

Comment:
This will allow us to ignore untracked files as the `git diff --name-only HEAD` command will return a list of tracked modified files that are either staged or unstaged

```suggestion
    if ((await git.diff(['--name-only', 'HEAD'])).length > 0) {
```

Reply:
The problem is that even if files are untracked, they could be imported by other files which _are_ tracked, leading to unreliable results.

Reply:
It happens, but I think it's rare that people create new files but forget to stage them. In those cases, CI will catch it. I know that we want to move more processing to the client, but I think this is an ok compromise that will only result in very few unnecessary CI runs.

Reply:
I just pulled down the latest changes in your PR branch and can now see that it has been changed to warning (sorry I missed that previously). I think this is also a good compromise. It lets the user know that there's an uncertainty in the test but allows the user to continue if they wish.